traitcuration/traits/datasources/ols.py

```diff
@@ -0,0 +1,36 @@
+import requests
+
+
+BASE_URL = "https://www.ebi.ac.uk/ols/api/"
+
+
+def make_ols_query(term_uri, ontology):
+    """
+    Takes in a term URI (or IRI as referenced in OLS documentation) and the ontology id to search against. Returns a
+    dictionary for that term with the fields 'label', 'obo_id' which is the CURIE and 'status'.
+    """
+    results = requests.get(
+        f"{BASE_URL}ontologies/{ontology}/terms?iri={term_uri}").json()
+    print(f"{BASE_URL}ontologies/{ontology}/terms?iri={term_uri}")
+    if 'error' in results:
+        return None
```

Comment:
Do you know in which cases it returns an 'error'? Is it only for missing terms, or can there be other scenarios?

Reply:
Yeah I have given that some thought. The only cases for errors I have found are '404 not found', the most common one for missing terms, and network error. Should I handle network errors any differently?

Reply:
Yes, network errors definitely have to be handled separately, as those "errors" are two really different cases:

- Term is (really) not found in EFO → requires import
- Term is "not found" in EFO due to network error → request should be retried, otherwise we will either get an incorrect status (when querying OLS) or we will miss some suggestions (when querying ZOOMA)

I suggest you handle requests in the following manner:
* For network errors, catch-and-retry with `@retry` decorator (see example [here](https://github.com/EBIvariation/vep-mapping-pipeline/blob/eee47471312cbcf08955acdd68d76851c5ffa5b4/vep_mapping_pipeline/consequence_mapping.py#L54)). If for a given entity we have exceeded the maximum number of retries, the entire update process should be halted and an error raised.
  + By the way, in production, all update operations (from ClinVar or from ZOOMA) will have to be started as transactions and rolled back in case of any exception, with the error being logged. I have created a future issue for this: https://github.com/EBIvariation/trait-curation/issues/52
* For _expected_ HTTP error codes, like a term being not found in EFO, verify that the error is of the right type: check the code and possibly also the contents of the response body (if applicable)
* For all other _unexpected conditions_ (like HTTP error codes which are not 404) raise a log an exception

Reply:
> Regarding the error handling, just to clarify that I am proceeding as I should, I am first checking if I get a 404, in which case I return `None` as the result of the query.

Yes, this is fine, as 404 is an expected error in this case.

> In any other case, I am using `response.raise_for_status()` to raise potential exceptions. Should I let `retry` catch unexpected HTTP errors and make the query again (this is what is done in the example you provided) or should I have it catch only network errors?

Hmm. Yes, good question. On the one hand, everything but 200 or 404 is not expected from OLS. On the other hand, sometimes the server can have a hiccup and return an unusual code _sporadically,_ which probably shouldn't fail the entire process right away.

So yes, let's use the `retry` as specified in my example, where it catches both network errors from `requests.get()/post()` _and_ unusual sporadic HTTP responses from `raise_for_status()`. Thank you for noticing and asking about this!